cpp/src/communicator/metadata_payload_exchange.cpp

```diff
@@ -0,0 +1,303 @@
+        );
+
+        // Send data immediately after metadata (if any)
+        if (message->data() != nullptr && message->data()->size > 0) {
+            auto data_buffer = message->release_data();
+            RAPIDSMPF_EXPECTS(data_buffer, "No data buffer available");
+
+            fire_and_forget_.push_back(
+                comm_->send(std::move(data_buffer), dst, gpu_data_tag_)
+            );
+        }
+    }
+
+    statistics_->add_duration_stat("comms-interface-send-messages", Clock::now() - t0);
+}
+
+std::vector<std::unique_ptr<MetadataPayloadExchange::Message>>
+TagMetadataPayloadExchange::receive_messages(
+    std::function<std::unique_ptr<Buffer>(std::size_t)> allocate_buffer_fn
+) {
+    auto const t0 = Clock::now();
+
+    // Process all phases of the communication protocol
+    receive_metadata();
+    setup_data_receives(allocate_buffer_fn);
+    auto completed_messages = complete_data_transfers();
+    cleanup_completed_operations();
+
+    statistics_->add_duration_stat("comms-interface-receive-messages", Clock::now() - t0);
+
+    return completed_messages;
+}
+
+bool TagMetadataPayloadExchange::is_idle() const {
+    return fire_and_forget_.empty() && incoming_messages_.empty()
+           && in_transit_messages_.empty() && in_transit_futures_.empty();
+}
+
+void TagMetadataPayloadExchange::receive_metadata() {
+    auto& log = comm_->logger();
+    auto const t0 = Clock::now();
+
+    while (true) {
+        auto const [msg, src] = comm_->recv_any(metadata_tag_);
+        if (!msg) {
+            break;
+        }
+
+        // Unpack metadata: [message_id][payload_size][original_metadata]
+        RAPIDSMPF_EXPECTS(
+            msg->size() >= sizeof(std::uint64_t) + sizeof(std::size_t),
+            "Truncated metadata"
+        );
+
+        std::size_t offset = 0;
+
+        // Extract message ID, payload size, and original metadata
+        std::uint64_t message_id;
+        std::memcpy(&message_id, msg->data() + offset, sizeof(std::uint64_t));
+        offset += sizeof(std::uint64_t);
+
+        std::size_t payload_size;
+        std::memcpy(&payload_size, msg->data() + offset, sizeof(std::size_t));
+        offset += sizeof(std::size_t);
+
+        std::vector<std::uint8_t> original_metadata(
+            msg->begin() + static_cast<std::ptrdiff_t>(offset), msg->end()
+        );
+
+        auto message = std::make_unique<MetadataPayloadExchange::Message>(
+            src, std::move(original_metadata), nullptr
+        );
+
+        log.trace("recv_any from ", src, " (message_id=", message_id, ")");
+        incoming_messages_.emplace(
+            src, TagMessage(std::move(message), message_id, payload_size)
+        );
+    }
+
+    statistics_->add_duration_stat("comms-interface-receive-metadata", Clock::now() - t0);
+}
+
+void TagMetadataPayloadExchange::setup_data_receives(
+    std::function<std::unique_ptr<Buffer>(std::size_t)> allocate_buffer_fn
```

Comment:
```suggestion
    std::function<std::unique_ptr<Buffer>(std::size_t)> const & allocate_buffer_fn
```

Reply:
Done in https://github.com/rapidsai/rapidsmpf/pull/437/commits/04cd976d3a459e76e007c4bcd8a265c8c6b44890